components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -2649,6 +2665,144 @@ void JsonRpcService::OnGetSPLTokenAccountBalance(
+    std::move(callback).Run(
+        "", mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  auto account_info_metadata_callback =
+      base::BindOnce(&JsonRpcService::OnGetSolanaAccountInfoMetaplex,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  // Call getAccountInfo for the metadata PDA
+  auto account_info_response_callback = base::BindOnce(
+      &JsonRpcService::OnGetSolanaAccountInfo, weak_ptr_factory_.GetWeakPtr(),
+      std::move(account_info_metadata_callback));
+  RequestInternal(solana::getAccountInfo(*associated_metadata_account), true,
+                  network_urls_[mojom::CoinType::SOL],
+                  std::move(account_info_response_callback));
+}
+
+void JsonRpcService::OnGetSolanaAccountInfoMetaplex(
+    GetMetaplexMetadataCallback callback,
+    absl::optional<SolanaAccountInfo> account_info,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess || !account_info) {
+    std::move(callback).Run(
+        "", mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  absl::optional<std::vector<uint8_t>> metadata =
+      base::Base64Decode((*account_info).data);
+  if (!metadata) {
+    std::move(callback).Run(
+        "", mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+  }
+
+  absl::optional<GURL> url =
+      solana_ins_data_decoder::DecodeMetadataUri(*metadata);
+  if (!url) {
+    std::move(callback).Run(
+        "", mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+  }
+
+  FetchTokenMetadata(
+      *url, mojom::CoinType::SOL,
+      base::BindOnce(&JsonRpcService::CompleteGetTokenMetadataSol,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void JsonRpcService::FetchTokenMetadata(
+    GURL url,
+    mojom::CoinType coin,
+    GetTokenMetadataIntermediateCallback callback) {
+  mojom::ProviderErrorUnionPtr error;
+  // Obtain JSON from the URL depending on the scheme.
+  // IPFS, HTTPS, and data URIs are supported.
+  // IPFS and HTTPS URIs require an additional request to fetch the metadata.
+  std::string metadata_json;
+  std::string scheme = url.scheme();
+#if BUILDFLAG(ENABLE_IPFS)
+  if (scheme != url::kDataScheme && scheme != url::kHttpsScheme &&
+      scheme != ipfs::kIPFSScheme) {
+#else
+  if (scheme != url::kDataScheme && scheme != url::kHttpsScheme) {
+#endif
+    if (coin == mojom::CoinType::SOL) {
+      error = mojom::ProviderErrorUnion::NewSolanaProviderError(
+          mojom::SolanaProviderError::kInternalError);
+    } else {
+      error = mojom::ProviderErrorUnion::NewProviderError(
+          mojom::ProviderError::kMethodNotSupported);
+    }
```

Comment:
In order to reuse the existing Ethereum code that parses the URI and fetches it IPFS/HTTPS or parses it as a data URI, I had to pass a CoinType parameter to the new `FetchTokenMetadata` function and then the function uses that parameter to either call the callback with a ProviderError or SolanaProviderError for the union.  An intermediate callback is used to unwrap the ProviderUnion and call the frontend's callback with the raw *ProviderError type.

This is kind of ugly, but I didn't see a better way that didn't require a fairly significant refactor.  Is there a better way?

Reply:
resolved by using integers for the error type and casting at the end